contrib/pkg/clusterpool/clusterpool.go

```diff
@@ -0,0 +1,419 @@
+			accessKeyID, secretAccessKey, err := awsutils.GetAWSCreds(o.CredsFile, defaultCredsFilePath)
+			if err != nil {
+				log.WithError(err).Error("Failed to get AWS credentials")
+				return nil, err
+			}
+			builder.CloudBuilder = &clusterresource.AWSCloudBuilder{
+				AccessKeyID:     accessKeyID,
+				SecretAccessKey: secretAccessKey,
+				Region:          o.Region,
+			}
+		case cloudAzure:
+			creds, err := azureutils.GetCreds(o.CredsFile)
+			if err != nil {
+				log.WithError(err).Error("Failed to read in Azure credentials")
+				return nil, err
+			}
+			builder.CloudBuilder = &clusterresource.AzureCloudBuilder{
+				ServicePrincipal:            creds,
+				BaseDomainResourceGroupName: o.AzureBaseDomainResourceGroupName,
+				Region:                      o.Region,
+			}
+		case cloudGCP:
+			creds, err := gcputils.GetCreds(o.CredsFile)
+			if err != nil {
+				log.WithError(err).Error("Failed to get GCP credentials")
+				return nil, err
+			}
+			builder.CloudBuilder = &clusterresource.GCPCloudBuilder{
+				ServiceAccount: creds,
+				Region:         o.Region,
+			}
+		}
+	}
+
+	imageSet, err := o.configureImages()
+	if err != nil {
+		return nil, err
+	}
+
+	result, err := o.build(builder)
+	if err != nil {
+		return nil, err
+	}
+
+	if imageSet != nil {
+		result = append(result, imageSet)
+	}
+	if len(pullSecret) != 0 {
+		result = append(result, builder.GeneratePullSecretSecret())
+	}
+
+	return result, nil
+}
+
+func (o *ClusterPoolOptions) configureImages() (*hivev1.ClusterImageSet, error) {
+	if len(o.ClusterImageSet) > 0 {
+		return nil, nil
+	}
+
+	if o.ReleaseImage == "" {
+		if o.ReleaseImageSource == "" {
+			return nil, fmt.Errorf("specify either a release image or a release image source")
+		}
+		var err error
+		o.ReleaseImage, err = utils.DetermineReleaseImageFromSource(o.ReleaseImageSource)
+		if err != nil {
+			return nil, fmt.Errorf("cannot determine release image: %v", err)
+		}
+	}
+	imageSet := &hivev1.ClusterImageSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("%s-imageset", o.Name),
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ClusterImageSet",
+			APIVersion: hivev1.SchemeGroupVersion.String(),
+		},
+		Spec: hivev1.ClusterImageSetSpec{
+			ReleaseImage: o.ReleaseImage,
+		},
+	}
+	o.ClusterImageSet = imageSet.Name
+	return imageSet, nil
+}
+
+func (o *ClusterPoolOptions) build(builder *clusterresource.Builder) ([]runtime.Object, error) {
+
+	var allObjects []runtime.Object
+
+	allObjects = append(allObjects, o.generateClusterPool())
+
+	if o.createCloudSecret {
+		cloudCredsSecret := builder.CloudBuilder.GenerateCredentialsSecret(builder)
+		if cloudCredsSecret != nil {
+			allObjects = append(allObjects, cloudCredsSecret)
+		}
+	}
+
+	return allObjects, nil
+}
+
+func (o *ClusterPoolOptions) getPullSecretSecretName() string {
+	return fmt.Sprintf("%s-pull-secret", o.Name)
+}
+
+func (o *ClusterPoolOptions) addClusterPoolPlatform(cp *hivev1.ClusterPool) {
```

Comment:
The code in this function is a duplicate of what is done by the CloudBuilder `addClusterDeploymentPlatform` function. Can you change that function to return a `*hivev1.Platform` instead of setting the platform on a cluster deployment?

Reply:
I don't follow what you are saying. I am suggesting that the code in `addClusterDeploymentPlatform` that creates a `hivev1.Platform` object be extracted out of that function so that it can be used by `addClusterDeploymentPlatform` and `addClusterPoolPlatform`. I don't understand what `Spec` has to do with that.

Reply:
... [Ah now I get it](https://media.giphy.com/media/TJawtKM6OCKkvwCIqX/giphy.gif) :disappointed: